airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+import traceback
+from abc import ABC
+from typing import Any, Dict, List, Mapping, Optional, Tuple, Type
+
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.file_based.availability_strategy import AbstractFileBasedAvailabilityStrategy
+from airbyte_cdk.sources.file_based.discovery_policy import AbstractDiscoveryPolicy, DefaultDiscoveryPolicy
+from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader
+from airbyte_cdk.sources.file_based.file_types import AvroParser, CsvParser, JsonlParser, ParquetParser
+from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
+from airbyte_cdk.sources.file_based.remote_file import FileType
+from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream, DefaultFileBasedStream
+
+default_parsers = {
+    FileType.Avro: AvroParser(),
+    FileType.Csv: CsvParser(),
+    FileType.Jsonl: JsonlParser(),
+    FileType.Parquet: ParquetParser(),
+}
+
+
+class FileBasedSource(AbstractSource, ABC):
+    """
+    All file-based sources must provide a `stream_reader`.
+    """
+
+    def __init__(
+        self,
+        stream_reader: AbstractFileBasedStreamReader,
+        availability_strategy: AbstractFileBasedAvailabilityStrategy,
+        discovery_policy: AbstractDiscoveryPolicy = DefaultDiscoveryPolicy(),
+        parsers: Dict[FileType, FileTypeParser] = None,
+        stream_cls: Type[AbstractFileBasedStream] = DefaultFileBasedStream,
```

Comment:
I keep getting hung up on this parameter. It feels a weird to me that we need to pass the type of FileBasedStream here because it makes it add another layer of obfuscation about what type it will actually be when we build the streams in the below `streams()` method. 

As in, let's say we have an S3 source that inherits FileBasedSource(). When we first `__init__()` the S3 file source, we also need to pass in what type of stream to be creating when first building the source during runtime. Whereas, if the S3 file source were to just reimplement the `streams()` method, it is statically specified in code exactly what type of stream it will be creating. It's readable in code either as the parameter or in the streams() method, but just feels a bit more complicated unless we needed the type to be dynamic at runtime or something.


Maybe if you cold provide a small code snippet of what it looks like to initialize a new file source that needs construct a new `AbstractFileBasedStream` that might help me understand the reason why we need to pass it in this way.

Reply:
I think I agree with Brian. It feels circuitous to pass the class when the dev can init and return it directly from `streams` (that direction also makes it easier for the dev to keep). It potentially complicates the story for how we init those streams, but I think that's solvable

Separately, I really second @brianjlai 's suggestion about keeping a "pet implementation" of a file-based connector going just as a demo alongside this project. It shouldn't be its own module, it can just be a package inside the CDK that you remove once this project is done. But I think it helps a lot if as you are designing these abstractions, you're seeing exactly what the user experience is like. I tried doing something like this in #27297 with the stripe connector (I just added a `demo` package in the CDK to test the UX as I go along)

Reply:
oh derp good point on inmemory -- i missed that